### PR TITLE
added support for fuzzy search via  let g:asyncomplete_smart_completion=1

### PR DESCRIPTION
**Pull request by prabirshrestha**
- [x] Supports vim8 with `has('lua')` and `exists('##TextChangedP')`
- [x] Supports neovim with  `has('nvim-0.2.2')` and `exists('##TextChangedP')`. neovim 0.2.2 already comes with lua support inbuilt. Currently it is throwing some error.

For now I will not expose the function to provide custom smart completion i.e. `s:filter_completion_items_lua(prefix, matches)` but should be fairly simple in the future.

Please give it a try. Currently it uses lua for library for smart completion so if anyone has a better algorithm I'm very open to replace it. Please give it a try, Will leave this PR open for few days and try it at real code in work before I merge this.

### CODE REVIEW

**Comment by prabirshrestha** [comment]
Neovim dev is failing with the following error.

image

File: autoload/asyncomplete.vim [hunk 1]
```diff
@@ -262,13 +264,15 @@ function! s:remote_refresh(ctx, force) abort
...
-        call asyncomplete#log('core', 's:remote_refresh', l:matchpos, a:ctx)
+        call asyncomplete#log('core', 's:remote_refresh', l:name, l:matchpos, a:ctx)
 
         let l:typed_len = l:endpos - l:startpos
         if l:typed_len == 1
             call add(l:sources_to_notify, l:name)
         elseif has_key(s:matches, l:name) && s:matches[l:name]['refresh']
             call add(l:sources_to_notify, l:name)
+        elseif s:supports_smart_completion()
+            call s:python_refresh_completions(a:ctx)
```
Comment (prabirshrestha) [inline autoload/asyncomplete.vim]: Queue in next tick for better performance.

**Comment by prabirshrestha** [comment]
@justinmk any reason why this wouldn't work in neovim? I'm using nvim v0.3.0-1289-g671840b9c dev on windows. Is there another api that I should be using?

```lua
local prefix = vim.eval('a:prefix')
```

**Comment by justinmk** [comment]
@prabirshrestha `vim.eval` is a "sugar" interface that we didn't implement, which is why we still return `0` for `has('lua')`: because didn't (yet) implement the full `if_lua` Vim legacy API.

Instead you can use `vim.api.nvim_eval()`.

**Comment by justinmk** [comment]
> But how can I add it to vim list and return it.

The Nvim Lua `vim.api` is a one-to-one match with the Nvim RPC API (`:help api`). So for setting VimL variables, there is `nvim_set_var()`. But that only works for global variables because script-local/function-local scope doesn't really make sense for RPC calls.

Regardless of Vim/Nvim it would make the most sense to me to write everything in Lua and only return data at the boundaries, rather than sprinkling `lua << EOF` here and there. Even with Vim's if_lua ability to transparently mutate VimL objects, this uses Lua "userdata" and I'm not sure that's optimal.  Instead do the work on Lua objects, then return the result.

The general pattern then is to use `luaeval()` to call Lua fuctions and use the magic `_A` Lua global to "marshal" VimL objects to Lua. (`_A` is documented at `:help luaeval()`, but it's a bit obtuse, so see the example below.)

From your sample code: 

```vim
lua << EOF
function filter_completion_items(prefix, matches)
  table.insert(matches, 'hello world')
  return matches
end
EOF
function! Filter_completion_items_lua(prefix, matches) abort
  " The 2nd arg to luaeval() can be a list so you can pass multiple args and
  " access them via _A[i].
  return luaeval('filter_completion_items(_A[1], _A[2])', [a:prefix, a:matches])
endfunction

echo Filter_completion_items_lua('XX', ['abc', '123', 'cdef', 'egaf awef'])
```

BTW where did `tmpmatches:add()` come from? Use `table.insert()` to append to Lua tables.

**Comment by prabirshrestha** [comment]
@justinmk ` let l:tmpmatches = []` is a local vimscript variable so was trying to call `call add(l:tmpmatches, 'hello')`. I will try your suggestion. Is it possible to create a list in lua and then return it as a list in vim instead?

**Comment by justinmk** [comment]
> Is it possible to create a list in lua and then return it as a list in vim instead?

The code I gave above shows that any Lua object (except lambdas) can be returned and is automatically marshaled to a VimL object.

**Comment by justinmk** [comment]
FWIW I added some extra hints to the Lua doc in https://github.com/neovim/neovim/pull/8679 . 

> Is it possible to create a list in lua and then return it as a list in vim instead?

```
lua << EOF
function create_list_lua()
  local l1 = { 'a,', 'b,', 'c', }
  local l2 = { 'X,', 'Y,', 'Z', }
  vim.api.nvim_set_var('l_from_lua', l2)
  return l1
end
function set_viml_global_from_lua()
  local l = { 'a,', 'b,', 'c', }
  return l
end
EOF
function! Create_list_viml() abort
  let l1 = luaeval('create_list_lua()')
  let l2 = g:l_from_lua
  return [l1, l2]
endfunction

echo Create_list_viml()
```

prints:

```
[['a,', 'b,', 'c'], ['X,', 'Y,', 'Z']]
```

**Comment by prabirshrestha** [comment]
@justinmk Update the PR based of your suggestion. Moved most of the logic to lua file so it is cleaner and now works on both vim8 and neovim. I did need to add a conversion function for vim8.

```vim
local is_neovim = vim['api'] ~= nil and vim.api['nvim_eval'] ~= nil

if is_neovim then
    function to_vim_list(obj)
        return obj
    end
else
    function to_vim_list(obj)
        return vim.list(obj)
    end
end
```

I like the fact that neovim autoconverts to respective vim type but to make plugin authors life easier would be good to have `nvim.list()`, `nvim.dict()` method supported.

* What is the best way to detect neovim in lua?
* Is there a way to find the script path of the lua file or require a relative lua file. Not an expert in lua so I'm currently creating a global var from vimscript that is accessible to lua.

```vim
let s:asyncomplete_folder = fnamemodify(expand('<sfile>:p:h') . '/../', ':p:h')

function! s:init_lua() abort
    exec 'lua asyncomplete_folder="' . s:asyncomplete_folder . '"'
    exec 'luafile ' . s:asyncomplete_folder . '/lua/asyncomplete.lua'
endfunction
```

Then in luafile.

```lua
local fts = dofile(asyncomplete_folder .. '/lua/fts_fuzzy_match.lua')
```

File: lua/asyncomplete.lua [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+function dump(o)
+   if type(o) == 'table' then
+      local s = '{ '
+      for k,v in pairs(o) do
+         if type(k) ~= 'number' then k = '"'..k..'"' end
+         s = s .. '['..k..'] = ' .. asyncomplete.dump(v) .. ','
+      end
+      return s .. '} '
+   else
```
[the remark this answers is not in the document]
Comment (prabirshrestha) [reply lua/asyncomplete.lua]: Updated. Is there a function in neovim to dump table and non tables with the same function? Currently this is what I'm doing.

```lua
    function dump(o)
        if type(o) == 'table' then
            return vim.api.nvim_call_function('string', o)
        else
            return tostring(o)
        end
    end
```

[the remark this answers is not in the document]
Comment (justinmk) [reply lua/asyncomplete.lua]: @prabirshrestha  It shouldn't matter if it's a table, `vim.api.nvim_call_function('string', o)` should work on anything.

**Pull request merged**